### PR TITLE
WFLY-16100, WFLY-16056 to add build compatible extension support

### PR DESCRIPTION
**Pull request by starksm64**
Signed-off-by: Scott M Stark <starksm64@gmail.com>
More work on https://issues.redhat.com/browse/WFLY-16100 that adds missing build compatible extension integration and dependencies. There is a problem with scanning deployments with BuildCompatibleExtensions as seen in this CDI test failure:

org.jboss.cdi.tck.tests.build.compatible.extensions.syntheticBeanWithLookup.SyntheticBeanWithLookupTest
```
org.jboss.weld.exceptions.AmbiguousResolutionException: WELD-001335: Ambiguous dependencies for type MyDependentBean with qualifiers 
 Possible dependencies: 
  - Managed Bean [class org.jboss.cdi.tck.tests.build.compatible.extensions.syntheticBeanWithLookup.MyDependentBean] with qualifiers [@Any @Default],
  - Managed Bean [class org.jboss.cdi.tck.tests.build.compatible.extensions.syntheticBeanWithLookup.MyDependentBean] with qualifiers [@Any @Default]
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.bean.builtin.InstanceImpl.checkBeanResolved(InstanceImpl.java:254)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.bean.builtin.InstanceImpl.get(InstanceImpl.java:113)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.cdi.tck.tests.build.compatible.extensions.syntheticBeanWithLookup.MyPojoCreator.create(MyPojoCreator.java:16)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.cdi.tck.tests.build.compatible.extensions.syntheticBeanWithLookup.MyPojoCreator.create(MyPojoCreator.java:9)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.lite.extension.translator.LiteExtensionTranslator.lambda$synthesis$0(LiteExtensionTranslator.java:157)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.bootstrap.events.configurator.BeanConfiguratorImpl$CreateCallback.create(BeanConfiguratorImpl.java:372)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.bootstrap.events.configurator.BeanConfiguratorImpl$ImmutableBean.create(BeanConfiguratorImpl.java:511)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.contexts.unbound.DependentContextImpl.get(DependentContextImpl.java:64)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:675)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.bean.builtin.InstanceImpl.getBeanInstance(InstanceImpl.java:262)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.bean.builtin.InstanceImpl.lambda$getHandle$0(InstanceImpl.java:202)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.util.LazyValueHolder$1.computeValue(LazyValueHolder.java:32)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.util.LazyValueHolder.get(LazyValueHolder.java:46)
	at org.jboss.weld.core@5.0.0.CR2//org.jboss.weld.bean.builtin.InstanceImpl$HandlerImpl.get(InstanceImpl.java:380)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.cdi.tck.tests.build.compatible.extensions.syntheticBeanWithLookup.SyntheticBeanWithLookupTest.test(SyntheticBeanWithLookupTest.java:41)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.testng.Arquillian$3.invoke(Arquillian.java:146)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.container.test.impl.execution.LocalTestExecuter.execute(LocalTestExecuter.java:57)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:103)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:90)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:133)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:105)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.EventImpl.fire(EventImpl.java:62)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.container.test.impl.execution.ContainerTestExecuter.execute(ContainerTestExecuter.java:36)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:103)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:90)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.test.impl.TestContextHandler.createTestContext(TestContextHandler.java:116)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.test.impl.TestContextHandler.createClassContext(TestContextHandler.java:83)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.test.impl.TestContextHandler.createSuiteContext(TestContextHandler.java:69)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:133)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.test.impl.EventTestRunnerAdaptor.test(EventTestRunnerAdaptor.java:139)
	at deployment.SyntheticBeanWithLookupTest355982c2ff53323b8c1b8eb3ad186685e835f45.war//org.jboss.arquillian.testng.Arquillian.run(Arquillian.java:138)
```

@manovotn it seems like the deployment is being treated as both one to be scanned for annotated classes and beans registered via the BCE.

### CODE REVIEW

File: weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/BuildCompatibleExtensionProcessor.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                        .tryRegisterExtension(loadClass(LITE_EXTENSION_TRANSLATOR, classLoader),
+                                              deploymentUnit);
+                log.debugf("Registered %s for %s", LITE_EXTENSION_TRANSLATOR, deploymentUnit.getName());
+            }
+        } finally {
+            setContextClassLoader(currentClassLoader);
+        }
+    }
+
+    private void setContextClassLoader(ClassLoader classLoader) {
```
Comment (bstansberry) [inline weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/BuildCompatibleExtensionProcessor.java:114]: WildFlySecurityManager.setCurrentContextClassLoaderPrivileged can handle the TCCL stuff.

Comment (starksm64) [reply weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/BuildCompatibleExtensionProcessor.java]: I have switched to use that.

**Comment by manovotn** [comment]
I think I know what's the cause of failures - the TCK archive is considered to be a bean archive and the bean definition is picked up. This is wrong because, by definition, an archive with CDI extension (any of the two) and no beans.xml is not a bean archive.

I might know where to fix this (`BeanArchiveProcessor`) and I also think we don't need a separate DPU to process BCE. Instead, we can add this to `WeldPortableExtensionProcessor` (and rename it) as it does very much the same thing.

The changes I mean are captured here - https://github.com/manovotn/wildfly/tree/weldLiteExtensionTranslator
However, I am getting unsatisfied dependency in the test you mentioned, I am trying to figure out why...

**Comment by manovotn** [comment]
> Ok, we can close this PR then.

Ok, I will open a new one with my changes later today and will mention you there.

**Comment by starksm64** [comment]
This is the https://github.com/jakartaredhat/wf-core-tck-runner.git runner project I am working on to test certifying a configuration of WildFly against the Jakarta EE 10 Core Profile TCK. The https://github.com/jakartaredhat/wf-core-tck-runner/blob/main/src/test/resources/tck-lite-suite.xml file is the subset of the CDI TCK I was working on to restrict the tests to CDI Lite features.

**Comment by bstansberry** [comment]
Closing in favor of #15474

**Pull request closed without merging**